sway-core/src/semantic_analysis/namespace/trait_map.rs

```diff
@@ -9,7 +9,7 @@ type TraitName = CallPath;
 /// Map of function name to [TyFunctionDeclaration](ty::TyFunctionDeclaration)
 type TraitMethods = im::HashMap<String, ty::TyFunctionDeclaration>;
 /// Map of trait name and type to [TraitMethods].
-type TraitImpls = im::HashMap<(TraitName, TypeId), TraitMethods>;
+type TraitImpls = im::Vector<((TraitName, TypeId), TraitMethods)>;
```

Comment:
`im` also exposes an [`OrdMap` collection](https://docs.rs/im/latest/im/ordmap/struct.OrdMap.html) which might be closer to the original behaviour - would that simplify this change (e.g. the `extend` implementation at least)?

Reply:
Hmm, I don't think so ? The `insert` method on `OrdMap` is copy-on-write, which is unnecessary for this application. https://docs.rs/im/latest/im/ordmap/struct.OrdMap.html#method.insert

Reply:
I mean that semantically, the `OrdMap` is much closer to the original `HashMap` behaviour, but is ordered and deterministic, and would mean we don't need any of the changes that were made to `extend` below.

> The insert method on OrdMap is copy-on-write, which is unnecessary for this application.

Note that a copy only occurs if the data is shared with another instance, (i.e. if the `OrdMap` was previously cloned elsewhere) and only the shared parts are copied.